docs/Threading.md

```diff
@@ -0,0 +1,167 @@
+The function [pulls the next thread from the run queue](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L234) and then [calls `_lthread_resume`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L242).
+
+The [`_lthread_resume`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L453) function does most of the work for running a thread.
+This sets up the initial thread state if required and then calls [`_switch`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L482), a short assembly routine that saves all registers that are preserved across function calls and restores the same set of registers from a buffer.
+The `_switch` routine [restores the stack pointer](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L147) of the thread being switched to and pushes the new instruction pointer value onto the new stack.
+
+If this thread is a new thread, its instruction pointer was [set during thread initialisation](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L540).
+If this thread has previously yielded, the instruction pointer is the [return address for the `_switch` call that switched away from the lthread](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L134).
+
+Note that `_switch` does *not* switch between two arbitrary lthreads.
+One of the threads involved in a `_switch` call is always the scheduler.
+The `_switch` call in `_lthread_resume` switches to another thread, the call [in `_lthread_yield_cb`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L340) and [in `_lthread_yield`](https://github.com/lsds/sgx-lkl/blob/47a5f0e718badfa85694a9de6222af41d9bfbb84/src/sched/lthread.c#L346) switch back to the scheduler.
+
+
+Locking
+-------
+
+There are two primitives for building locks in the lthreads implementation.
+Ticket locks are spinlocks, futexes provide a fast path using atomic operations but then sleep.
+
+### Ticked locks
+
+Ticket locks are defined and documented in [src/include/enclave/ticketlock.h](src/include/enclave/ticketlock.h).
+Acquiring a ticket lock will spin in a compare and swap (CAS) loop.
+If an lthread yields while holding a ticket lock, there is a chance of deadlock because the next thread to try to acquire the lock will spin in a CAS loop and not yield.
+This may, in turn, prevent the thread holding the lock from being able to run.
+
+Ticket locks are used to implement some parts of the lthread scheduler.
+This means that there is no way for them to be made to yield in their spin loop.
```

Comment:
I think the last 2 sentences are confusing. I'm not sure what you are trying to convey here.

Reply:
Better?